sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,843 @@
+        self.X_, y = check_X_y(X, y, order='F')
+        check_classification_targets(y)
+
+        # Store the appearing classes and the class index for each sample
+        self.labels_, self.y_ = np.unique(y, return_inverse=True)
+        self.classes_ = np.arange(len(self.labels_))
+
+        # Check that the number of neighbors is achievable for all classes
+        self.n_neighbors_ = self.check_n_neighbors(self.y_)
+        # TODO: Notify superclass KNeighborsClassifier that n_neighbors
+        # might have changed to n_neighbors_
+        # super().set_params(n_neighbors=self.n_neighbors_)
+
+        # Initialize transformer
+        self.L_, self.n_features_out_ = self._init_transformer()
+
+        # Prepare for saving if needed
+        if self.save is not None:
+            save_dir, save_file = os.path.split(self.save)
+            if save_dir != '' and not os.path.exists(save_dir):
+                os.mkdir(save_dir)
+            save_file = self.save + '_' + str(self.n_funcalls_)
+            np.save(save_file, self.L_)
+
+        # Find target neighbors (fixed)
+        self.targets_ = self._select_target_neighbors()
+
+        # Compute gradient component of target neighbors (constant)
+        self.grad_static_ = self._compute_grad_static()
+
+        # Initialize number of optimizer iterations and objective funcalls
+        self.n_iter_ = 0
+        self.n_funcalls_ = 0
+
+        # Set some default values for the optimizer
+        max_corr = 100
+        max_funcalls = 2*self.max_iter
+
+        # Call optimizer
+        try:
+            L, loss, info = optimize.fmin_l_bfgs_b(func=self._loss_grad,
+                                                   x0=self.L_,
+                                                   m=max_corr,
+                                                   pgtol=self.tol,
+                                                   maxiter=self.max_iter,
+                                                   maxfun=max_funcalls,
+                                                   disp=self.disp,
+                                                   callback=self._cb)
+        except TypeError:
+            # Type Error caused in old versions of SciPy because of no
+            # maxiter argument (<= 0.9).
+            self.L_ = np.asfortranarray(self.L_)
+            try:
+                L, loss, info = optimize.fmin_l_bfgs_b(func=self._loss_grad,
+                                                       x0=self.L_,
+                                                       m=max_corr,
+                                                       maxfun=max_funcalls,
+                                                       disp=self.disp,
+                                                       pgtol=self.tol)
+            except Exception:
+                print('lbfgs does not work as expected in this version of '
+                      'Scipy. Probably it is too old.')
+                return self
```

Comment:
I was raising the exception in the previous commit. I just used this to make sure this is the only case the travis tests fail. I will look into the old scipy documentation and see if I can do something with `sklearn.utils.fixes` like you suggested. From a first look, this old version of scipy uses an older version of the optimizer too which would give worse solutions.